[1LP][WIPTEST] Implement Openstack create VM.

ldjebran opened the pull request:
```python
rom wrapanapi import OpenstackSystem

os_system = OpenstackSystem(
    username=<hidden>,
    password=<hidden>,
    domain_name=<hidden>,
    auth_url=<hidden>,
    tenant=<hidden>,
    app_name="iqe-insights-satellite", app_version="0.1"
)


print("  - test existing open stack instance >>>")
instance = os_system.get_instance("iqe-satellite-test-dlezz-uoaxjidnrg")
print(instance.name)
print(instance.uuid)
print(instance.exists)
print(instance.state)
print(instance.ip)

print(" - test creating a new open stack instance >>>")
new_ins = os_system.create_vm(
    image="rhel-7.6-server-x86_64-released",
    name="iqe-satellite-test-dlezz-124",
    flavor="m1.small",
    key_name=<hidden>,
    network="25ec4907-36fd-7035-b8d5-b797246330f2",
)
# network modified , showing here only the format
assert new_ins.exists
print(new_ins.name)
print(new_ins.uuid)
print(new_ins.exists)
print(new_ins.state)
print(new_ins.ip)
print('delete instance ...')
new_ins.cleanup()
assert not new_ins.exists
```
# test output
```
projects/iqe_insights_satellite/scripts/twrapenapi.py
  - test existing openstack instance >>>
iqe-satellite-test-dlezz-uoaxjidnrg
18ccf865-d41e-4e3f-b740-78463c5525fc
True
VmState.RUNNING
15.2.150.67
 - test creating a new openstack instance >>>
Unmapped VM state 'BUILD' received from system, mapped to 'VmState.UNKNOWN'
iqe-satellite-test-dlezz-124
75fd181d-1482-47f5-bbce-c80293bd21b1
True
VmState.RUNNING
15.2.149.11
delete instance ...
# note the ips was modified in this output .
```
Was able to test it only for api version 3, our upshift service is not configured to allow listing all tenants and has no floating ip configuration.

mshriver commented on wrapanapi/systems/openstack.py:
```diff
@@ -130,17 +132,26 @@ def _get_state(self):
+                if nic['OS-EXT-IPS:type'] == ip_type:
                     return str(nic['addr'])
 
+    @property
```
Please see #386 for updates to IP lookup.

ldjebran replied:
Thanks, will rebase than and test it.

ldjebran replied:
looked at the code in that PR, think for openstack nothing should change for this PR, as all_ips will return all the available netwoks ips, but here we need external ip, especially when assiging a floating ip, all_ips will return internal and external (as I guess) [ip property should be external ip (fixed or floating)].

mshriver commented on wrapanapi/systems/openstack.py:
```diff
@@ -164,6 +175,21 @@ def rename(self, new_name):
         self.refresh()  # update raw
 
+    def auto_assign_floating_ip(self, timeout=120):
+        """Auto assign a floating IP to this VM instance.
```
I would much rather see this as an update to `assign_floating_ip`, making `floating_ip_pool` argument optional.

ldjebran replied:
implemented

mshriver commented on wrapanapi/systems/openstack.py:
```diff
@@ -824,6 +905,13 @@ def get_vm(self, name=None, id=None, ip=None):
         return matches[0]
 
+    def get_instance(self, vm_name_or_id):
+        """Return the virtual machine instance by it's id or name"""
```
I think this will be a bit confusing/misleading when `get_vm` exists.  `get_vm` takes advantage of `find_vm`, and the fact that multiple may exist with the same name.  What's the advantage here, just using the connection object to look up instead of using an API client object?

ldjebran replied:
The problem here is that get_vm try to get what we may have no access to when listing all tenants and in this case we do not what to filter any thing, we know the exact name of the vm or it's id and  we what to get it's wrapanapi instance (perhaps we can rename it to to get_vm_instance as the original get_vm return a munch) 
here is what we get when trying to get the vm with get_vm function
```python
os_system.get_vm(name="iqe-satellite-test-dlezz-uoaxjidnrg")
Traceback (most recent call last):
  File "/home/dlezz/.pyenv/versions/insights/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3291, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-6-9072f2dc6070>", line 1, in <module>
    os_system.get_vm(name="iqe-satellite-test-dlezz-uoaxjidnrg")
  File "/home/dlezz/projects/wrapanapi/wrapanapi/systems/openstack.py", line 901, in get_vm
    matches = self.find_vms(**kwargs)
  File "/home/dlezz/projects/wrapanapi/wrapanapi/systems/openstack.py", line 865, in find_vms
    instances = self.list_vms()
  File "/home/dlezz/projects/wrapanapi/wrapanapi/systems/openstack.py", line 834, in list_vms
    instances = self._generic_paginator(call)
  File "/home/dlezz/projects/wrapanapi/wrapanapi/systems/openstack.py", line 814, in _generic_paginator
    temp_list = f()
  File "/home/dlezz/.pyenv/versions/insights/lib/python3.6/site-packages/novaclient/v2/servers.py", line 861, in list
    "servers")
  File "/home/dlezz/.pyenv/versions/insights/lib/python3.6/site-packages/novaclient/base.py", line 254, in _list
    resp, body = self.api.client.get(url)
  File "/home/dlezz/.pyenv/versions/insights/lib/python3.6/site-packages/keystoneauth1/adapter.py", line 375, in get
    return self.request(url, 'GET', **kwargs)
  File "/home/dlezz/projects/wrapanapi/wrapanapi/systems/openstack.py", line 56, in _request_timeout_handler
    return SessionClient.request(self, url, method, **kwargs)
  File "/home/dlezz/.pyenv/versions/insights/lib/python3.6/site-packages/novaclient/client.py", line 80, in request
    raise exceptions.from_response(resp, body, url, method)
novaclient.exceptions.Forbidden: Policy doesn't allow os_compute_api:servers:detail:get_all_tenants to be performed. (HTTP 403) (Request-ID: req-6d1e9bd3-75e3-4405-b337-c21db450eb75)

```

ldjebran replied:
Please also note that I wanted to use the openstack well documented api for getting a concrete  server that is via the connection, in our case getting the connection object is cheap as it's via the already created session.

ldjebran replied:
but we can also modify get_vm function to not search for the vm if we do not want to:

```python
def get_vm(self, name=None, id=None, ip=None, search=True):
    # search is True to not break compatibility
    if search is False:
         name_or_id:
              # prefere id or name
              name_or_id = id if id else name
              return self.connection.get_server(name_or_id) 
```
and of course remove that get_instance
what do you think?

ldjebran replied:
found an other approach 
```python
    def list_vms(self, filter_tenants=True):
        call = partial(self.api.servers.list, True, {'all_tenants': filter_tenants})

def find_vms(self, name=None, id=None, ip=None, filter_tenants=True):
    ...
    instances = self.list_vms(filter_tenants=filter_tenants)

def get_vm(self, name=None, id=None, ip=None, filter_tenants=True):
    kwargs = {'name': name, 'id': id, 'ip': ip, 'filter_tenants': filter_tenants}
    ...
```

when testing
```python
vm = os_system.get_vm(name="my-rhel8", filter_tenants=False)
vm
Out[5]: <wrapanapi.systems.openstack.OpenstackInstance system=<OpenstackSystem> raw=<novaclient.v2.servers.Server>, kwargs['uuid']='bf7cac9e-2abb-46e0-b7fb-1bb15508afa6'>

```
what do you think?

ldjebran replied:
function removed

mshriver commented on wrapanapi/systems/openstack.py:
```diff
@@ -713,8 +775,27 @@ def remove_tenant(self, tenant_name):
+    def create_vm(self, wait_for_state=VmState.RUNNING, wait_for_state_timeout=360,
+                  wait_for_state_delay=15, assign_floating_ip=False, floating_ip_pool=None,
+                  assign_floating_ip_timeout=120, **kwargs):
+        vm = self.connection.create_server(**kwargs)
```
I see you're taking advantage of the connection object to call create_server, and passing it all the kwargs. Any reason you would not get an `OpenstackImage` object, and then use its `deploy` method to deploy the instance from that image?

ldjebran replied:
Thanks for the suggestion and Ok if this is the standard way to provision a VM, made some tests and seems also working.
```python
In[28]: image = os_system.get_template(name="rhel-7.6-server-x86_64-released")
In[29]: ins = image.deploy(
   ...:     "iqe-satellite-test-dlezz-124",
   ...:     flavor_name="m1.small",
   ...:     key_name=<hidden>,
   ...:     network_name=<hidden>
   ...: )
Unmapped VM state 'BUILD' received from system, mapped to 'VmState.UNKNOWN'
Unmapped VM state 'BUILD' received from system, mapped to 'VmState.UNKNOWN'
Unmapped VM state 'BUILD' received from system, mapped to 'VmState.UNKNOWN'
In[30]: ins.ip
Out[30]: '20.15.145.135'
# IP <modified>
In[31]: ins.name
Out[31]: 'iqe-satellite-test-dlezz-124'
In[32]: ins.exists
Out[32]: True
In[33]: ins.state
Out[33]: 'VmState.RUNNING'
In[34]: ins.uuid
Out[34]: '4942117f-a1fe-4757-bb85-b22ee2aea293'
In[35]: ins.cleanup()
Out[35]: True
In[36]: ins.exists
Out[36]: False
```
will update the code than.

ldjebran commented:
@mshriver many thanks for the review and updated the code taking into account your suggestions.

mshriver commented on wrapanapi/systems/openstack.py:
```diff
@@ -173,6 +173,17 @@ def all_ips(self):
+        floating_ip = self.floating_ip
+        if floating_ip:
+            return floating_ip
+        return self.get_ip_by_type('fixed')
```
This is a change to the behavior of this property - I think it should still return `None` if there is no floating IP found. Right now the caller expects this `ip` property to return a 'reachable' address, not an internal/fixed address, and `None` if that is not found.

With this change, the caller will now not know whether the returned address is the floating address.

I would suggest continuing to return `None` here if a floating IP isn't found, and the caller can then resort to directly calling `get_ip_by_type` directly if they would like to proceed with a fixed address.

ldjebran replied:
removed and let default ip property to be equal to floating ip, as some tests are expecting this value to be floating ip.

mshriver commented on wrapanapi/systems/openstack.py:
```diff
@@ -153,15 +155,13 @@ def _get_state(self):
 
-    @property
-    def ip(self):
+    def get_ip_by_type(self, ip_type):
```
I think we should provide a default for ip_type here, and maybe include some OpenstackSystem constants for what types are available. That way callers can use this method in a more predictable way, `osp_instance.get_ip_by_type(osp_system.FIXED_ADDR)`

ldjebran replied:
included default fixed value and 
```
NIC_IP_TYPE_FIXED = "fixed"
NIC_IP_TYPE_FLOATING = "floating"
```

mshriver commented on wrapanapi/systems/openstack.py:
```diff
@@ -173,6 +173,17 @@ def all_ips(self):
         return [ip for nets in self.raw.networks.values() for ip in nets]
 
+    @property
+    def floating_ip(self):
```
Consider having explicit methods for `floating_ip` and `fixed_ip`.

ldjebran replied:
done

ldjebran commented:
@mshriver @izapolsk changes done
```python
from wrapanapi import OpenstackSystem
from wrapanapi.systems.openstack import OpenstackImage

def print_instance_info(ins):
    print('Name:', ins.name)
    print('UUID:', ins.uuid)
    print('Exists:', ins.exists)
    print('State:', ins.state)
    print('IP:', ins.ip)
    print('Floating IP:', ins.floating_ip)
    print('Fixed IP:', ins.fixed_ip)

os_system = OpenstackSystem(
    username=<hidden>,
    password=<hidden>,
    domain_name=<hidden>,
    auth_url=<hidden>,
    tenant="hidden",
    app_name="iqe-insights-satellite", app_version="0.1"
)

print("  - test existing open stack instance >>>")
instance = os_system.get_vm(name="jslave-rhel-75-drift-883", filter_tenants=False)
print_instance_info(instance)

new_ins = os_system.create_vm(
    "iqe-satellite-test-dlezz-124",
    image_name="rhel-7.6-server-x86_64-released",
    flavor_name="m1.small",
    key_name=<hidden>,
    network_name=<hidden>,
)
assert new_ins.exists
print_instance_info(new_ins)
print('delete instance ...')
new_ins.cleanup()
assert not new_ins.exists
```
results:
```bash
iqe_insights_satellite/scripts/twrapenapi.py
  - test existing open stack instance >>>
Name: jslave-rhel-75-drift-883
UUID: 4f8f15e1-d986-4933-9817-7d0248024bb8
Exists: True
State: VmState.RUNNING
IP: None
Floating IP: None
Fixed IP: 121.2.117.17  <modified>
 - test creating a new open stack instance >>>
Unmapped VM state 'BUILD' received from system, mapped to 'VmState.UNKNOWN'
Unmapped VM state 'BUILD' received from system, mapped to 'VmState.UNKNOWN'
Unmapped VM state 'BUILD' received from system, mapped to 'VmState.UNKNOWN'
Name: iqe-satellite-test-dlezz-124
UUID: b72463aa-9d4a-4160-883e-5ae0fc112955
Exists: True
State: VmState.RUNNING
IP: None
Floating IP: None
Fixed IP: 121.2.145.222 <modified>
delete instance ...

Process finished with exit code 0
```

ldjebran commented:
rebased, fixed and removed some code that was implemented when waiting for this PR to be merged.

ldjebran commented:
@mshriver @izapolsk is there any updates about this PR?

mshriver commented on wrapanapi/systems/openstack.py:
```diff
@@ -174,6 +177,18 @@ def all_ips(self):
+        return self.get_ip_by_type(NIC_IP_TYPE_FLOATING)
+
+    @property
+    def ip(self):
```
Thanks so much for structuring this in this explicit way!

izapolsk commented on wrapanapi/systems/openstack.py:
```diff
@@ -758,8 +801,35 @@ def remove_tenant(self, tenant_name):
+
+        Args:
+            name (str): The name of the VM instance to create.
+            image_name (str): From witch image name to deploy the new VM instance.
```
burning her is the best what we can do in this case :)

mshriver commented on wrapanapi/systems/openstack.py:
```diff
@@ -27,6 +27,7 @@
 from novaclient import exceptions as os_exceptions
 from novaclient.client import SessionClient
 from novaclient.v2.floating_ips import FloatingIP
+from openstack.connection import Connection
```
We don't directly require openstack SDK, we're only getting it through our dependency on python-ironicclient.

You need to update the requirements file to specify a version of openstacksdk.

Running this with current requirements, we're getting openstacksdk==0.27.0 installed, and the latest is 0.32.0.

```[WARNING] [py.warnings] ./.cfme_venv/lib64/python3.7/site-packages/openstack/_meta/connection.py:121: ImportWarning: Could not import data-processing service description: No module named 'openstack.data_processing'
  ImportWarning)
```

This version, along with our other package versions, is resulting in the above import error - along with a few dozen others from connection module.

Please evaluate the package version needs for python-*client and python-openstacksdk or openstacksdk requirement.

mshriver replied:
@ldjebran I'm willing to accept the OpenstackSystem.create_vm method, as its part of the original design  as an abstract method.

The above needs to be resolved though for the PR to be accepted

izapolsk commented on wrapanapi/systems/openstack.py:
```diff
@@ -758,8 +801,35 @@ def remove_tenant(self, tenant_name):
 
-    def create_vm(self):
-        raise NotImplementedError('create_vm not implemented.')
+    def create_vm(self, name, image_name=None, image_id=None, assign_floating_ip=False,
```
I'm wondering what it is for. we already have deploy method which helps us to deploy instance from image. I guess this one just duplicates functionality.

izapolsk commented on wrapanapi/systems/openstack.py:
```diff
@@ -758,8 +801,35 @@ def remove_tenant(self, tenant_name):
 
-    def create_vm(self):
-        raise NotImplementedError('create_vm not implemented.')
+    def create_vm(self, name, image_name=None, image_id=None, assign_floating_ip=False,
```
honestly, I'd put this in image.deploy and make this create_vm just alias to image.deploy()

ldjebran replied:
The purpose of this function is a bit different from just image deploy, it court circuit the deployment of the image from image name or image id and assign floating ip is needed, to have a ready accessible vm. But personnally prefer the more general connection.create_server that has more options (name or id of different components) usage in this regards that was implemented in earlier commits.

izapolsk replied:
actually image.deploy() does the same. I understand your confusion. This is called so due to historical reasons. either way, I don't make you move this code there.

mshriver replied:
I think we can implement for the rest of the system modules by having create_vm implemented to pull the template object and deploy it.  I think the implementation here is acceptable, and if we would prefer the functionality in the deploy method we can move it there when standardizing implementation of create_vm.

ldjebran commented:
@mshriver sorry for the delay, 

added the dependency, the same as osc-lib used by python-ironicclient
https://github.com/openstack/python-ironicclient/blob/master/requirements.txt
https://github.com/openstack/osc-lib/blob/master/requirements.txt
